tests/auto/qtpromise/qpromise/tst_convert.cpp

```diff
@@ -0,0 +1,257 @@
+
+        Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<QString>>::value));
+
+        QCOMPARE(waitForValue(p, QString{}), QString{"42"});
+        QVERIFY(p.isFulfilled());
+    }
+
+    // Conversion of a non-Qt type via QVariant.
+    // https://doc.qt.io/qt-5/qmetatype.html#registerConverter
+    {
+        auto p = QtPromise::resolve(Foo{42}).convert<QString>();
+
+        Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<QString>>::value));
+
+        QCOMPARE(waitForValue(p, QString{}), QString{"Foo{42}"});
+        QVERIFY(p.isFulfilled());
+    }
+}
+
+void tst_qpromise_convert::fulfillTAsVoid()
+{
+    auto p = QtPromise::resolve(42).convert<void>();
+
+    Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<void>>::value));
+
+    QCOMPARE(waitForValue(p, -1, 42), 42);
+    QVERIFY(p.isFulfilled());
+}
+
+void tst_qpromise_convert::fulfillTAsQVariant()
+{
+    // Primitive type to QVariant.
+    {
+        auto p = QtPromise::resolve(42).convert<QVariant>();
+
+        Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<QVariant>>::value));
+
+        QCOMPARE(waitForValue(p, QVariant{}), QVariant{42});
+        QVERIFY(p.isFulfilled());
+    }
+
+    // Non-Qt user-defined type to QVariant.
+    {
+        auto p = QtPromise::resolve(Foo{42}).convert<QVariant>();
+
+        Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<QVariant>>::value));
+
+        QVariant value = waitForValue(p, QVariant{});
+        QCOMPARE(value, QVariant::fromValue(Foo{42}));
+        QCOMPARE(value.value<Foo>().m_foo, 42);
+        QVERIFY(p.isFulfilled());
+    }
+}
+
+void tst_qpromise_convert::fulfillQVariantAsU()
+{
+    // Test whether a directly stored value can be extracted.
+    {
+        auto p = QtPromise::resolve(QVariant{42}).convert<int>();
+
+        Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<int>>::value));
+
+        QCOMPARE(waitForValue(p, -1), 42);
+        QVERIFY(p.isFulfilled());
+    }
+
+    // Test automatic conversion from string performed by QVariant.
+    // https://doc.qt.io/qt-5/qvariant.html#toInt
+    {
+        auto p = QtPromise::resolve(QVariant{"42"}).convert<int>();
+
+        Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<int>>::value));
+
+        QCOMPARE(waitForValue(p, -1), 42);
+        QVERIFY(p.isFulfilled());
+    }
+
+    // Non-Qt user-defined type
+    {
+        auto p = QtPromise::resolve(QVariant::fromValue(Foo{42})).convert<Foo>();
+
+        Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<Foo>>::value));
+
+        QCOMPARE(waitForValue(p, Foo{}), Foo{42});
+        QVERIFY(p.isFulfilled());
+    }
+}
+
+void tst_qpromise_convert::fulfillQVariantAsVoid()
+{
+    auto p = QtPromise::resolve(QVariant{42}).convert<void>();
+
+    Q_STATIC_ASSERT((std::is_same<decltype(p), QPromise<void>>::value));
+
+    QCOMPARE(waitForValue(p, -1, 42), 42);
+    QVERIFY(p.isFulfilled());
+}
+
+void tst_qpromise_convert::rejectUnconvertibleQVariant()
```

Comment:
We should probably test rejecting for non Qt types, for example: `std::vector` to `int`?

Reply:
Good point, I've added a test for it